utils/unittest/entity_test.go

```diff
@@ -0,0 +1,162 @@
+package unittest
+
+import (
+	"testing"
+
+	"github.com/onflow/crypto"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/exp/slices"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type IntList []uint32
+
+func (l IntList) ID() flow.Identifier {
+	return flow.MakeID(l)
+}
+
+type StructWithNilFields struct {
+	Identities flow.IdentitySkeletonList
+	Index      map[flow.Identifier]uint32
+	QC         *flow.QuorumCertificate
+}
+
+func (e *StructWithNilFields) ID() flow.Identifier {
+	type pair struct {
+		ID    flow.Identifier
+		Index uint32
+	}
+	var pairs []pair
+	for id, index := range e.Index {
+		pairs = append(pairs, pair{id, index})
+	}
+	slices.SortFunc(pairs, func(a, b pair) int {
+		return flow.IdentifierCanonical(a.ID, b.ID)
+	})
+	return flow.MakeID(struct {
+		Identities flow.IdentitySkeletonList
+		Index      []pair
+		QcID       flow.Identifier
+	}{
+		Identities: e.Identities,
+		Index:      pairs,
+		QcID:       e.QC.ID(),
+	})
+}
+
+type StructWithUnsupportedFlowField struct {
+	field flow.IdentitySkeleton
+}
+
+func (e *StructWithUnsupportedFlowField) ID() flow.Identifier {
+	return flow.MakeID(e)
+}
+
+// MalleableEntityStruct is a struct that is malleable because its ID method does not cover all of its fields.
+type MalleableEntityStruct struct {
+	Identities flow.IdentitySkeletonList
+	QC         *flow.QuorumCertificate
+	Signature  crypto.Signature
+}
+
+// ID returns the hash of the entity in a way that does not cover all of its fields.
+func (e *MalleableEntityStruct) ID() flow.Identifier {
+	return flow.MakeID(struct {
+		Identities flow.IdentitySkeletonList
+		QcID       flow.Identifier
+	}{
+		Identities: e.Identities,
+		QcID:       e.QC.ID(),
+	})
+}
+
+// TestRequireEntityNonMalleable tests the behavior of MalleabilityChecker with different types of entities ensuring
+// it correctly handles the supported types and returns an error when the entity is malleable, or it cannot perform the check.
+func TestRequireEntityNonMalleable(t *testing.T) {
```

Comment:
I think we could add one more struct type and test here to test and document usage of the malleability:"optional" tag

Reply:
[`27718b3` (#7069)](https://github.com/onflow/flow-go/pull/7069/commits/27718b39413da52b9b25179acee2357af9c2ba0c)